src/Microsoft.ML.Experimental/DataLoadSave/Database/DatabaseLoader.cs

```diff
@@ -290,31 +313,102 @@ public sealed class Options
+
+            public Segment(int min, int lim, bool forceVector)
+            {
+                Contracts.Assert(0 <= min & min < lim);
+                Min = min;
+                Lim = lim;
+                ForceVector = forceVector;
+            }
+        }
+
         /// <summary>
         /// Information for an output column.
         /// </summary>
         private sealed class ColInfo
         {
             public readonly string Name;
-            public readonly int? SourceIndex;
             public readonly DataViewType ColType;
+            public readonly Segment[] Segments;
 
-            public ColInfo(string name, int? sourceIndex, DataViewType colType)
+            // BaseSize is the sum of the sizes of segments.
+            public readonly int SizeBase;
+
+            private ColInfo(string name, DataViewType colType, Segment[] segs, int sizeBase)
             {
                 Contracts.AssertNonEmpty(name);
-                Contracts.Assert(!sourceIndex.HasValue || sourceIndex >= 0);
-                Contracts.AssertValue(colType);
+                Contracts.AssertValueOrNull(segs);
+                Contracts.Assert(sizeBase > 0);
 
                 Name = name;
-                SourceIndex = sourceIndex;
+                Contracts.Assert(colType.GetItemType().GetRawKind() != 0);
                 ColType = colType;
+                Segments = segs;
+                SizeBase = sizeBase;
+            }
+
+            public static ColInfo Create(string name, PrimitiveDataViewType itemType, Segment[] segs, bool user)
+            {
+                Contracts.AssertNonEmpty(name);
+                Contracts.AssertValue(itemType);
+                Contracts.AssertValueOrNull(segs);
+
+                int size = 0;
+                DataViewType type = itemType;
+
+                if (segs != null)
+                {
+                    var order = Utils.GetIdentityPermutation(segs.Length);
+                    Array.Sort(order, (x, y) => segs[x].Min.CompareTo(segs[y].Min));
+
+                    // Check that the segments are disjoint.
+                    for (int i = 1; i < order.Length; i++)
+                    {
+                        int a = order[i - 1];
+                        int b = order[i];
+                        Contracts.Assert(segs[a].Min <= segs[b].Min);
+                        if (segs[a].Lim > segs[b].Min)
+                        {
+                            throw user ?
+                                Contracts.ExceptUserArg(nameof(Column.Source), "Intervals specified for column '{0}' overlap", name) :
+                                Contracts.ExceptDecode("Intervals specified for column '{0}' overlap", name);
+                        }
+                    }
+
+                    // Note: since we know that the segments don't overlap, we're guaranteed that
+                    // the sum of their sizes doesn't overflow.
+                    for (int i = 0; i < segs.Length; i++)
```

Comment:
What happens now when a `DatabaseLoader` is saved to a model file, when there are no Segments you write a `0` in for the segment length.

Then, when that `DatabaseLoader` gets loaded from the model file, it creates a new array of length `0` (in the Bindings constructor that takes a ModelLoadContext), which gets passed into this method. Nothing will increment the `int size`, so it will remain `0`, and pass into the constructor above, which will trigger an assert.